test/strategy/FluidStrategy.t.sol

```diff
@@ -0,0 +1,434 @@
+        vm.stopPrank();
+    }
+
+    function test_RevertWhen_NotOwner() public useActor(0) {
+        vm.expectRevert("Ownable: caller is not the owner");
+        strategy.setPTokenAddress(ASSET, P_TOKEN);
+    }
+
+    function test_RevertWhen_InvalidPToken() public useKnownActor(USDS_OWNER) {
+        address INVALID_P_TOKEN = 0xe0C97480CA7BDb33B2CD9810cC7f103188de4383;
+        vm.expectRevert();
+        strategy.setPTokenAddress(ASSET, INVALID_P_TOKEN);
+    }
+
+    function test_RevertWhen_InvalidPToken_anotherAsset() public useKnownActor(USDS_OWNER) {
+        address ANOTHER_ASSET_PTOKEN = 0xbE3860FD4c3facDf8ad57Aa8c1A36D6dc4390a49;
+        vm.expectRevert(abi.encodeWithSelector(InvalidAssetLpPair.selector, ASSET, ANOTHER_ASSET_PTOKEN));
+        strategy.setPTokenAddress(ASSET, ANOTHER_ASSET_PTOKEN);
+    }
+
+    function test_SetPTokenAddress() public useKnownActor(USDS_OWNER) {
+        assertEq(strategy.assetToPToken(ASSET), address(0));
+
+        vm.expectEmit(true, false, false, false);
+        emit PTokenAdded(address(ASSET), address(P_TOKEN));
+        strategy.setPTokenAddress(ASSET, P_TOKEN);
+
+        assertEq(strategy.assetToPToken(ASSET), P_TOKEN);
+        assertTrue(strategy.supportsCollateral(ASSET));
+    }
+
+    function test_RevertWhen_DuplicateAsset() public useKnownActor(USDS_OWNER) {
+        strategy.setPTokenAddress(ASSET, P_TOKEN);
+        vm.expectRevert(abi.encodeWithSelector(PTokenAlreadySet.selector, ASSET, P_TOKEN));
+        strategy.setPTokenAddress(ASSET, P_TOKEN);
+    }
+}
+
+contract RemovePTokenTest is FluidStrategyTest {
+    function setUp() public override {
+        super.setUp();
+        vm.startPrank(USDS_OWNER);
+        _initializeStrategy();
+        _setAssetData();
+        vm.stopPrank();
+    }
+
+    function test_RevertWhen_NotOwner() public useActor(0) {
+        vm.expectRevert("Ownable: caller is not the owner");
+        strategy.removePToken(0);
+    }
+
+    function test_RevertWhen_InvalidId() public useKnownActor(USDS_OWNER) {
+        vm.expectRevert(abi.encodeWithSelector(InvalidIndex.selector));
+        strategy.removePToken(5);
+    }
+
+    function test_RevertWhen_CollateralAllocated() public useKnownActor(USDS_OWNER) {
+        _deposit();
+        vm.expectRevert(abi.encodeWithSelector(CollateralAllocated.selector, ASSET));
+        strategy.removePToken(0);
+    }
+
+    function test_RemovePToken() public useKnownActor(USDS_OWNER) {
+        assertEq(strategy.assetToPToken(ASSET), P_TOKEN);
+        assertTrue(strategy.supportsCollateral(ASSET));
+
+        vm.expectEmit(true, false, false, false);
+        emit PTokenRemoved(ASSET, P_TOKEN);
+        strategy.removePToken(0);
+
+        (uint256 allocatedAmt) = strategy.allocatedAmount(ASSET);
+
+        assertEq(allocatedAmt, 0);
+        assertEq(strategy.assetToPToken(ASSET), address(0));
+        assertFalse(strategy.supportsCollateral(ASSET));
+    }
+}
+
+contract DepositTest is FluidStrategyTest {
```

Comment:
Add one test case where allocate is called from the vault and vault is calling deposit on strategy. Real world use case.

Reply:
Same for withdraw